File: Daily/PY/Leetcode496-下一个更大元素.py
from typing import List
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        lb = [0] * len(nums2)
        stk = []
        n,m = len(nums1),len(nums2)
        for i in range(m - 1,-1,-1):
            t = nums2[i]
            while stk and t >= stk[-1]: stk.pop()
            if not stk: lb[i] = -1
            else: lb[i] = stk[-1]
            stk.append(t)
        mp = dict()
        for i in range(m):
            mp[nums2[i]] = i
        ans = [0] * n
        for i in range(n):
            ans[i] = lb[mp[nums1[i]]]
        return ans

File: Leetcode/sbgo/Leetcode496-下一个更大元素.go
package sbgo

func nextGreaterElement(nums1 []int, nums2 []int) []int {
	n, m := len(nums1), len(nums2)
	lb := make([]int, m)
	stk := []int{}
	for i := m - 1; i >= 0; i-- {
		t := nums2[i]
		for len(stk) != 0 && stk[len(stk)-1] <= t {
			stk = stk[0 : len(stk)-1]
		}
		if len(stk) == 0 {
			lb[i] = -1
		} else {
			lb[i] = stk[len(stk)-1]
		}
		stk = append(stk, t)
	}
	ans := make([]int, n)
	mp := map[int]int{}
	for i := 0; i < m; i++ {
		mp[nums2[i]] = i
	}
	for i := 0; i < n; i++ {
		ans[i] = lb[mp[nums1[i]]]
	}
	return ans
}
